### PR TITLE
OME005: Add additional writer implementation details

### PR DESCRIPTION
**Pull request by rleigh-codelibre**
Update the proposal with detailed steps for implementing OME-TIFF pyramid writing, including references to work in progress so far.

### CODE REVIEW

File: OME005/index.md [hunk 1]
```diff
@@ -263,54 +264,103 @@ Follow-up work could include:
...
+the internal writer state will be inconsistent.  `setSeries` can't be
+called before `setId`, and so this approach is incompatible with the
+current writer semantics.
+
+Secondly, by setting the core metadata list:
+
+| IFormatWriter       | Description
+| ------------------- | -------------------------------------------------------
+| setCoreMetadataList | Set the core metadata list (all series and resolutions)
+| getCoreMetadataList | Set the core metadata list (all series and resolutions)
```
Comment (mtbc) [inline OME005/index.md]: get?

File: OME005/index.md [hunk 2]
```diff
@@ -349,6 +399,119 @@ required.  With the corresponding reader support, this would provide
...
+to loop over each resolution as well as each series and transfer all
+the resolution levels.  The `imageconverter-noflat` and
+`ometiff-pyramid-writer` branches implement most of the needed logic.
+
+## Outstanding questions
+
+- Do we need to store metadata to describe alignment offsets between
+  sub-resolution levels when the size difference between levels
+  results in pixels not aligned on pixel boundaries, for example with
+  non-power-of-two reductions.  How is this handled by the existing
```
Comment (mtbc) [inline OME005/index.md]: question mark

File: OME005/index.md [hunk 3]
```diff
@@ -349,6 +399,119 @@ required.  With the corresponding reader support, this would provide
...
+  pyramid file formats? (Question from 2018 OME annual meeting in
+  Dundee).
```
Comment (mtbc) [inline OME005/index.md]: period inside the ()

File: OME005/index.md (hunk 2 above)
Comment (rleigh-codelibre) [reply OME005/index.md]: There's a question mark at the end of the sentence on the following line.

**Comment by rleigh-codelibre** [comment]
@mtbc All your comments should have fixes now, with the exception of the question mark, because I already see one!

File: OME005/index.md (hunk 2 above)
Comment (mtbc) [reply OME005/index.md]: However this sentence too appears to be a question: it starts "Do we" and asks a different question from that which follows.

File: OME005/index.md [hunk 4]
```diff
@@ -349,6 +399,119 @@ required.  With the corresponding reader support, this would provide
...
+`ometiff-pyramid-writer` branches implement most of the needed logic.
+
+## Outstanding questions
+
+- Do we need to store metadata to describe alignment offsets between
+  sub-resolution levels when the size difference between levels
+  results in pixels not aligned on pixel boundaries, for example with
+  non-power-of-two reductions.  How is this handled by the existing
+  pyramid file formats? (Question from 2018 OME annual meeting in
+  Dundee.)
```
Comment (melissalinkert) [inline OME005/index.md]: We should also address whether or not there is any enforcement of the XY sizes of each resolution.  In particular, what happens if someone tries to write a pyramid of e.g. 1024x1024 => 512x512 => 200x200?

**Comment by sbesson** [comment]
Thanks very much for documenting the current status of the implementation and the next proposed steps. Pushed a commit for the two obvious typos listed above.

Some of the questions raised by @melissalinkert are important to address while we define the boundaries of this extension. In order to deliver a working writing implementation in the upcoming weeks, we might need to make some decisions about the range of behaviors and layouts we will commit to support e.g. enforcing some constraints on the downsampling factor between resolutions sounds to be in agreement with existing pyramid formats but also a restriction we could decide to lift in the future if we wished so.

Merging this to update the state of the public design. The next step will be to turn it into a branch against the model with the formal specification extension.

**Comment by rleigh-codelibre** [comment]
@sbesson When it does come to considering the writer API design, do bear in mind http://lists.openmicroscopy.org.uk/pipermail/ome-devel/2018-June/004260.html where the C# wrapper already added support for direct conversion of CoreMetadata to OMEXMLMetadata.  This round-tripping is what's needed here as well, and the fact that it's been independently done without any input from ourselves highlights its utility and necessity.  Whether it's `setCoreMetadata` in addition to `setMetadataRetrieve` or adding utility methods to achieve the same conversion in `MetadataTools`, round-tripping from one to the other is essential, and that will require proper subchannel sizes in CoreMetadata to round trip from Channel in the data model.

However, CoreMetadata is not something we really want to expose unless strictly necessary.  Having the writer support the whole set of `setXXX` methods might be worthwhile for direct support for image acquisition, as opposed to conversion.  Conversion is as simple as copying the metadata store or the core metadata list.  But acquisition may require more direct interactivity, and we should consider this also.  Specifying every series and dimension size up front is too restrictive, and relaxing this is primarily blocked by the `initialized` array in `FormatWriter` which should be pretty simple to adjust.  You can only write the series in order, so it could be dynamically set per series.

**Pull request merged**